### PR TITLE
 Add getSinglerRef function for temporary retrieval of reference datasets

### PR DESCRIPTION
**Pull request by j-andrews7**
As mentioned in #1, this provides a way to retrieve reference datasets from Github. 

Not quite sure I'm using the `envir` parameter for `load` in quite the way you wanted, but this avoids muddying the `GlobalEnv` and the `eval` tomfoolery I had previously.

I'm willing to add a note about it to the vignette if you'd like.

### CODE REVIEW

File: R/helper.R [hunk 1]
```diff
@@ -0,0 +1,43 @@
+#' Retrieve SingleR reference datasets from Github
+#'
+#' \code{getSinglerRef} retrieves \code{SingleR} reference datasets from Github. 
+#'
+#' @param dataset String referring to SingleR reference dataset to retrieve.
```
Comment (LTLA) [inline R/helper.R]: There's no need to explicitly list the options here, if they are implicitly present in the argument list of the function (see my other comment). I would recommend moving a full description of the options into `@details` along with some plain-English details about what each one is. This information is probably in the old SingleR repo.

File: R/helper.R [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+#'
+#' @examples
+#'
+#' immgen <- getSinglerRef(dataset = "immgen")
+#' mouse.rnaseq <- getSinglerRef(dataset = "mouse.rnaseq")
+#'
+#' @export
+#' @importFrom curl curl_download
+#' 
+getSinglerRef <- function(dataset = "hpca") {
```
Comment (LTLA) [inline R/helper.R]: Move `opts` as the default `dataset=` argument and use `dataset <- match.arg(dataset)` to do the check.

File: R/helper.R [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+#'
+#' @author Jared Andrews
+#'
+#' @examples
+#'
+#' immgen <- getSinglerRef(dataset = "immgen")
+#' mouse.rnaseq <- getSinglerRef(dataset = "mouse.rnaseq")
+#'
+#' @export
+#' @importFrom curl curl_download
```
Comment (LTLA) [inline R/helper.R]: Switch the imports to pull the relevant functions out of _BiocFileCache_, and don't forget to add _BiocFileCache_ to the `Imports:` of the `DESCRIPTION`.

File: R/helper.R [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+	  }
+	  base.url <- "https://github.com/dviraran/SingleR/blob/master/data/"
+	  url.tail <- ".rda?raw=true"
+
+	  full.url <- paste0(base.url, dataset, url.tail)
+
+	  bfc <- BiocFileCache(ask=FALSE)
+	  ref <- bfcrpath(bfc, full.url)
+
+	  load(ref, envir = environment())
```
Comment (LTLA) [inline R/helper.R]: The way to do it is this:

```r
env <- new.env()
load(ref, envir=env)
ref.set <- get(dataset, envir=env)
```

File: R/helper.R [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+#'
+#' @param dataset String referring to SingleR reference dataset to retrieve. 
+#'     Options include:
+#'     \itemize{
+#'       \item "hpca"
+#'       \item "blueprint_encode"
+#'       \item "immgen"
+#'       \item "mouse.rnaseq"
+#'     }
+#' @return SingleR reference dataset.
```
Comment (LTLA) [inline R/helper.R]: This is a bit too vague; I believe the output is a list with various fields of interest? Have a look at `trainSingleR`'s `@return` field for the level of detail required; think of what a skillful newcomer would need to interpret the output.

File: R/helper.R [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+getSinglerRef <- function(dataset = "hpca") {
+	  opts <- c("hpca", "blueprint_encode", "immgen", "mouse.rnaseq")
+	  if (!(dataset %in% opts)) {
+		    stop("Invalid SingleR reference set. Valid options are 'hpca',", 
+			      "'blueprint_encode', 'immgen', or 'mouse.rnaseq'.")
+	  }
+	  base.url <- "https://github.com/dviraran/SingleR/blob/master/data/"
+	  url.tail <- ".rda?raw=true"
+
+	  full.url <- paste0(base.url, dataset, url.tail)
```
Comment (LTLA) [inline R/helper.R]: This is fine, but you could also consider the more elegant:

```r
full.url <- sprintf("https://github.com/dviraran/SingleR/blob/master/data/%s.rda?raw=true", dataset)
```

**Comment by LTLA** [comment]
Also, perhaps the name `getSinglerRef` is not ideal, for various reasons:

1.  It's already in the `SingleR` package, so there's no need to have `SingleR` in the name as well.
2.  `Singler` itself is misspelt to fit into the camelCase but this is a bit awkward.

Suggest maybe something like `getReferenceDatasets`.

> I'm willing to add a note about it to the vignette if you'd like.

Yes, the vignette has a stub usage section for the reference datasets. If there's one that works well with the pancreas datasets that I've currently got in there, that would be good to demonstrate.

**Comment by j-andrews7** [comment]
Latest commit should have addressed all of your comments, I think. My code in the vignette takes ~90 seconds, which I'm guessing pushes the vignette pretty close to the limit.

Let me know if you see any other issues.

**Comment by LTLA** [comment]
Looks good. Would it help to subset the data in the vignette? Or is the bottleneck not related to the number of input cells?

**Comment by j-andrews7** [comment]
I didn't do any real benchmarking or debugging to test, but my guess is
that it's because the reference set is 700+ samples. And using the
pre-defined `de.genes` is a no-go, as deriving the common set likely
invalidates many of them. And I didn't super relish going through the lists
of lists to snag only those remaining.

If it's necessary, I'm sure there are ways to speed it up (including
subsetting the data - probably the simplest solution). There are ~1800 in
that set; subsetting to 500 would probably cut it down to a more manageable
level.

**Comment by LTLA** [comment]
Good enough, I'll sort it out at some point.

**Comment by LTLA** [comment]
Oh, I should have told you to add yourself as a contributor. I can do that if that's fine with you.

**Pull request merged**